cluster/gce/gci/configure-helper.sh

```diff
@@ -1701,17 +1701,43 @@ function prepare-etcd-manifest {
   if [[ -n "${INITIAL_ETCD_CLUSTER_STATE:-}" ]]; then
     cluster_state="${INITIAL_ETCD_CLUSTER_STATE}"
   fi
+
+  # Configure mTLS for etcd peers.
   if [[ -n "${ETCD_CA_CERT:-}" && -n "${ETCD_PEER_KEY:-}" && -n "${ETCD_PEER_CERT:-}" ]]; then
     etcd_creds=" --peer-trusted-ca-file /etc/srv/kubernetes/etcd-ca.crt --peer-cert-file /etc/srv/kubernetes/etcd-peer.crt --peer-key-file /etc/srv/kubernetes/etcd-peer.key -peer-client-cert-auth "
     etcd_protocol="https"
   fi
 
-  # mTLS should only be enabled for etcd server but not etcd-events. if $1 suffix is empty, it's etcd server.
+  # In a zonal cluster, etcd_host_ip and host_ip are both the same as host VM's internal IP.
+  # In a regional cluster, a secondary alias IP range (usually 10.127.X.X) is allocated for etcd servers.
+  local host_ip=""
+  local etcd_addtional_listen_client_url=""
+  local etcd_addtional_listen_metrics_url=""
+  if [[ ${ETCD_LISTEN_ON_HOST_IP:-} == "true" ]] ; then
+    if [[ -n ${HOST_PRIMARY_IP:-} ]]; then
+      host_ip=${HOST_PRIMARY_IP}
+    else
+      host_ip=$(hostname -i)
+    fi
+  fi
+
+  # Configure mTLS for clients (e.g. kube-apiserver).
+  # mTLS should only be enabled for etcd server but not etcd-events. If $1 suffix is empty, it's etcd server.
   if [[ -z "${suffix}" && -n "${ETCD_APISERVER_CA_KEY:-}" && -n "${ETCD_APISERVER_CA_CERT:-}" && -n "${ETCD_APISERVER_SERVER_KEY:-}" && -n "${ETCD_APISERVER_SERVER_CERT:-}" && -n "${ETCD_APISERVER_CLIENT_KEY:-}" && -n "${ETCD_APISERVER_CLIENT_CERT:-}" ]]; then
     etcd_apiserver_creds=" --client-cert-auth --trusted-ca-file ${ETCD_APISERVER_CA_CERT_PATH} --cert-file ${ETCD_APISERVER_SERVER_CERT_PATH} --key-file ${ETCD_APISERVER_SERVER_KEY_PATH} "
     etcd_apiserver_protocol="https"
     etcd_livenessprobe_port="2382"
-    etcd_extra_args+=" --listen-metrics-urls=http://${ETCD_LISTEN_CLIENT_IP:-127.0.0.1}:${etcd_livenessprobe_port} "
+
+    # To not require mTLS for metrics collection.
+    if [[ ${ETCD_LISTEN_ON_HOST_IP:-} == "true" ]] ; then
+      etcd_addtional_listen_metrics_url=",http://${host_ip}:${etcd_livenessprobe_port}"
+    fi
+    etcd_extra_args+=" --listen-metrics-urls=http://${ETCD_LISTEN_CLIENT_IP:-127.0.0.1}:${etcd_livenessprobe_port}${etcd_addtional_listen_metrics_url} "
+  fi
+
+  # If etcd is configured to listen on host IP, an additional listen-client URL is added.
+  if [[ ${ETCD_LISTEN_ON_HOST_IP:-} == "true" ]] ; then
```

Comment:
It feels a little strange to see the same condition appear in two blocks that are so close. Any chance of putting them together?

Reply:
I had the same thought. But it turns out the logic here really requires two control blocks... The control block inside mTLS block above needs both conditions (ETCD_LISTEN_ON_HOST_IP and suffix has value). And combining the if (ETCD_LISTEN_ON_HOST_IP) logic will make the code less readable.